### ISSUE TITLE
Parse error with non-ascii characters

### ISSUE THREAD

**Issue by clawoflight** [libofx/libofx#60]
libofx stopped working for me with non-ascii characters.
Example error messages:

> LibOFX ERROR: OpenSP parser: otherError (misc parse error):
/tmp/libofxtmpoEFTsQ:48:39:E: non SGML character number 195
> (Above message occurred on Line 48, Column 40)
> LibOFX ERROR: OpenSP parser: otherError (misc parse error):
/tmp/libofxtmpoEFTsQ:48:40:E: non SGML character number 164
> (Above message occurred on Line 48, Column 41)

This can be reproduced by using `ofxdump` or by attempting to import the file in Skrooge, which I use for my finances.
I assume gnucash and kmymoney will have the same issue.

My distro is Arch Linux, where this started happening in ~March 2022; identical files could be imported previously.
The AppImage for skrooge also seems to produce this issue, as does my wife's OpenSuse Tumbleweed system.

Example file with .txt extension so github allows it: [ofx-broken.ofx.txt](https://github.com/libofx/libofx/files/8884237/ofx-broken.ofx.txt)

**Comment by cstim** [comment]
Which version of libofx and opensp is this, for completeness? Thanks!

**Comment by clawoflight** [comment]
libofx:     0.10.5
opensp: 1.5.2

**Comment by cstim** [comment]
Versions are most recent, thanks.

In the other bug, @benoitg explained here https://github.com/libofx/libofx/issues/20#issuecomment-412183565 that there are a few environment variables for opensp (as documented here http://openjade.sourceforge.net/doc/charset.htm ) which are set in lib/ofx_preproc.cpp, line 195ff. When changing SP_ENCODING=utf-8, the error message about the non-ascii character in line 48 changes compared to the original message (which is funny enough) but it is still an error:
`
LibOFX ERROR: OpenSP parser: otherError (misc parse error):
/tmp/libofxtmpNNHEFC:48:39:E: non SGML character number 228
`
That being said, it's not clear to me how to get the opensp library (from 2005, yay) to accept the utf-8 encoding here. And the very long function ofx_preproc.cpp also doesn't make it clear where to place the relevant conversions, oh well.

By the way the characters still appear in the parsed string, don't they?

**Comment by jralls** [comment]
@cstim I haven't looked at the OpenSP code in a couple of years and I'm fuzzy on it. Does it understand utf16? If so we could transcode to that instead of utf8 then transcode the OpenSP parse tree back to utf8 for processing in libofx.

**Comment by cstim** [comment]
@jralls The thing where I got stuck is that libofx in preproc does not apply iconv to the file if it is identified as XML, and it wasn't obvious how to get iconv to process the file in that case anyway. Maybe you or me find something later...

**Comment by cstim** [comment]
@clawoflight I can clearly reproduce the error message. However, is there an actual error afterwards? Because with `ofxdump`, the error message is shown but the parsed strings contain the utf-8 characters just fine. Or is it just me?

**Comment by clawoflight** [comment]
@cstim I'm at work now, but IIRC you are right, the final output is correct.
But `ofxdump` exits with 1, not 0, if the error is displayed, so maybe the error handlig is incorrect?
Since the import into skrooge also fails, whatever is going wrong is probably somewhere deep in the library and not restricted to CLI.

**Comment by benoitg** [comment]
Ok, I'm pretty sure I wrote the following comment in the code:

              /* Normally the following would be "xml".
              * Unfortunately, opensp's generic api will garble UTF-8 if this is
              * set to xml.  So we set any single byte encoding to avoid messing
              * up UTF-8.  Unfortunately this means that non-UTF-8 files will not
              * get properly translated.  We'd need to manually detect the
              * encoding in the XML header and convert the xml with iconv like we
              * do for SGML to work around the problem.  Most unfortunate. */

Now there are 3 problems.
1. Even if we did parse the encoding in the XML file, we'd still have OpenSP errors for UTF-8 files, as there are no safe encodings to convert to.  In fact I just noticed that this can ALSO happen for SGML files since we also disabled UTF-8 support for those, as it was broken on windows (if memory serves). 
1. The return code from ofxdump comes from deep in OpenSP.  Should'nt be as bad for a normal application, but they may have to regex the "non SGML character number" in their error handler if they don't want to ignore other errors.
1.  There were no clear warning message to explain what is happening.  I just fixed that in #ef48375c4f whichi should at least make this less baffling for people.  The message is: 

> LibOFX WARNING: ofx_proc_file(): OpenSP cannot process an UTF-8 XML file without garbling it.  This is worked around by forcing a single byte encoding.  If the file is indeed UTF-8, it should pass through unmolested, but you will likely get 'non SGML character number' errors, even though the output is correct.

**Comment by clawoflight** [comment]
Hi Benoit, thank you for your investigation!
The warning is a very helpful first step, since it will at least prevent all of this digging from being needed again ;)

2. Seems like the deal breaker that originally brought me here, since skrooge just fails when trying to import an OFX with non-ascii characters. How good is the error design in opensp? Would it be possible for libofx to catch a specific exception, ignore a specific return code, or similar?

Otherwise, maybe it's time to switch to a different xml parser? libxml2 and expat seem to be very common, and both have good C++ wrappers. I would be willing to help with the port :)

**Comment by benoitg** [comment]
OpenSP's error handling isn't very granular, but that's a bit immaterial.  We could definitely squelch the error, and just output a warning within libofx just by matching the string "non SGML character number" for "otherError", as OpenSP hasn't ben updated in well over a decade and the string is unlikely to change (and we are in bad workaround territory already).  We should not squelch it completely, as we don't know the original file was in UTF-8 to begin with.  A patch would be welcome.

Supporting libxml's SAX2 API has been considered since the begining of LibOFX.  The problem is that since libxml2 does not support SGML, and most OFX files in the wild are still SGML, it would be a lot of code duplication, while still having to support OpenSP for most files.  Unless we have a maintainer for it, nobody was ever willing to spend the time.  But if you are...

**Comment by cstim** [comment]
I agree with @benoitg here: As we are in workaround territory anyway, we can just as well add string matching for "non SGML character number", and if that is the case, don't treat this as a terminating error but rather like a warning.

**Comment by wrobelda** [comment]
This may be relevant: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=225375

**Comment by clawoflight** [comment]
On June 29, 2022 10:23:32 PM GMT+02:00, "Benoit Grégoire" ***@***.***> wrote:
>OpenSP's error handling isn't very granular, but that's a bit immaterial.  We could definitely squelch the error, and just output a warning within libofx just by matching the string "non SGML character number" for "otherError", as OpenSP hasn't ben updated in well over a decade and the string is unlikely to change (and we are in bad workaround territory already).  We should not squelch it completely, as we don't know the original file was in UTF-8 to begin with.  A patch would be welcome.
>
>Supporting libxml's SAX2 API has been considered since the begining of LibOFX.  The problem is that since libxml2 does not support SGML, and most OFX files in the wild are still SGML, it would be a lot of code duplication, while still having to support OpenSP for most files.  Unless we have a maintainer for it, nobody was ever willing to spend the time.  But if you are...
>

Then I would suggest that you to ahead and handle that error, if we are always in workarounds that seems worth it. And then I can do my taxes :)

When I'm back from vacation, I'll have a look at the code base and the SAX2 API and decide whether I want to start working on that. 
Thanks for the help!
